File: queue.js
// Queue is a FIFO (first in first out) data structure
class Queue {
  constructor(array = []) {
    this.array = array;
  }

  // O(1)
  peek() {
    if (this.isEmpty()) return null;
    return this.array[0];
  }

  // O(1)
  enqueue(value) {
    this.array.push(value);
  }

  // O(1)
  isEmpty() {
    return this.array.length === 0;
  }

  // O(1)
  dequeue() {
    return this.array.shift();
  }
}

module.exports = Queue;

File: queue.py
class Queue:
  def __init__(self):
    self.items = []
  
  def enqueue(self, val):
    self.items.append(val)
  
  def dequeue(self):
    if not self.is_empty():
      return self.items.pop(0)

  def is_empty(self):
    return len(self.items) == 0

  def peek(self):
    if not self.is_empty():
      return self.items[0]
# Time complexity:
# Operation         Analysis
# Enqueue           O(1)
# Dequeue           O(1)
# Peek              O(1)
